packages/ifpack2/src/Ifpack2_Zoltan2Partitioner_def.hpp

```diff
@@ -0,0 +1,111 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+//@HEADER
+*/
+
+#ifndef IFPACK2_ZOLTAN2_PARTITIONER_DEF_HPP
+#define IFPACK2_ZOLTAN2_PARTITIONER_DEF_HPP
+
+#if defined(HAVE_IFPACK2_ZOLTAN2)
+#include "Ifpack2_ConfigDefs.hpp"
+#include "Ifpack2_Zoltan2Partitioner_decl.hpp"
+
+namespace Ifpack2 {
+
+//==============================================================================
+// Constructor
+template<class GraphType>
+Zoltan2Partitioner<GraphType>::
+Zoltan2Partitioner (const Teuchos::RCP<const row_graph_type>& graph) :
+  zoltan2AlgoName_ ("phg"), OverlappingPartitioner<GraphType> (graph)
+{}
+
+
+template<class GraphType>
+Zoltan2Partitioner<GraphType>::~Zoltan2Partitioner() {}
+
+
+template<class GraphType>
+void
+Zoltan2Partitioner<GraphType>::
+setPartitionParameters (Teuchos::ParameterList& List) {
+  // Default is Parallel Hypergraph
+  zoltan2AlgoName_ = List.get<std::string>("zoltan2: algorithm", zoltan2AlgoName_);
+}
+
+template<class GraphType>
+void Zoltan2Partitioner<GraphType>::computePartitions()
+{
+  // Create an input adapter for the Tpetra matrix.
+  Zoltan2GraphAdapterType zoltan2_graph(this->Graph_);
+
+  // Specify partitioning parameters
+  Teuchos::ParameterList zoltan2_params;
+  zoltan2_params.set("partitioning_approach", "partition");
+  zoltan2_params.set("num_local_parts", this->NumLocalParts_);
+  if (zoltan2AlgoName_ == "parmetis") {
+    zoltan2_params.set("algorithm", "parmetis");
+    zoltan2_params.set("symmetrize_input", "transpose"); // not sure if this does anything, and may fail with non-symm graph
+    zoltan2_params.set("partitioning_objective", "minimize_cut_edge_weight");
+  } else {
+    zoltan2_params.set("algorithm", zoltan2AlgoName_);
+  }
+
+  // Create and solve partitioning problem
+  Zoltan2::PartitioningProblem<Zoltan2GraphAdapterType>
+    problem(&zoltan2_graph, &zoltan2_params, this->Graph_->getComm());
+  problem.solve();
```

Comment:
The adapter takes a `Tpetra::CrsGraph` and a communicator. So I'd assume that this is solving an MPI global partitioning problem.

Reply:
I think you are right. Even if you set the num_local_parts parameter, I believe Zoltan2 will solve the global partitioning problem, then assign parts to processes. In this use case, it is better to call Zoltan2 with the local graph on each process.

Reply:
I am trying to figure out a good way to get the local graph. `Ifpack2`'s `LocalFilter` seems to return the graph of the original matrix, and `Tpetra`'s `RowGraph` does not seem to have `getLocalGraph`? @jhux2

Reply:
In fact, `LocalFilter` indicates that `getGraph` doesn't do what doxygen says it does.
https://github.com/trilinos/Trilinos/blob/7c431d00fa6873e55b604b8c5fda358f3f050f21/packages/ifpack2/src/Ifpack2_LocalFilter_def.hpp#L288-L298

Reply:
In RILUK, the LocalFilter is converted to an actual localized CrsMatrix.   What type of matrix is coming in to this class?

Reply:
Thank you, @jhux2. I like to get the local part of `A_`, which is of type RowMatrix, before calling Zoltan2 from Block Relaxation, [here](https://github.com/iyamazaki/Trilinos/blob/cd99d2f07eb354e6f1a2656df88c7a95125c8a1d/packages/ifpack2/src/Ifpack2_BlockRelaxation_def.hpp#L658-L659). So, I should follow the same step, and then get the local graph?